modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java

```diff
@@ -0,0 +1,889 @@
+            this.pageStoreFactory = pageStoreFactory;
+        }
+
+        /** */
+        public PageStore createPartPageStore() throws IgniteCheckedException {
+            PageStore partPageStore;
+
+            defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadLock();
+            try {
+                partPageStore = pageStoreFactory.createPageStore(
+                    FLAG_DATA,
+                    () -> defragmentedPartTmpFile(workDir, partId).toPath(),
+                    partPagesAllocated::addAndGet
+                );
+            }
+            finally {
+                defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadUnlock();
+            }
+
+            partPageStore.sync();
+
+            DefragmentationPageReadWriteManager pageMgr = (DefragmentationPageReadWriteManager)partPageMemory.pageManager();
+
+            pageMgr.pageStoreMap().addPageStore(grpId, partId, partPageStore);
+
+            return partPageStore;
+        }
+
+        /** */
+        public PageStore createMappingPageStore() throws IgniteCheckedException {
+            PageStore mappingPageStore;
+
+            defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadLock();
+            try {
+                mappingPageStore = pageStoreFactory.createPageStore(
+                    FLAG_DATA,
+                    () -> defragmentedPartMappingFile(workDir, partId).toPath(),
+                    mappingPagesAllocated::addAndGet
+                );
+            }
+            finally {
+                defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadUnlock();
+            }
+
+            mappingPageStore.sync();
+
+            DefragmentationPageReadWriteManager partMgr = (DefragmentationPageReadWriteManager)mappingPageMemory.pageManager();
+
+            partMgr.pageStoreMap().addPageStore(grpId, partId, mappingPageStore);
+
+            return mappingPageStore;
+        }
+
+        /** */
+        public LinkMap createLinkMapTree(boolean initNew) throws IgniteCheckedException {
+            defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadLock();
+
+            //TODO Store link in meta page and remove META_PAGE_IDX constant?
+            try {
+                long mappingMetaPageId = initNew
+                    ? mappingPageMemory.allocatePage(grpId, partId, FLAG_DATA)
+                    : PageIdUtils.pageId(partId, FLAG_DATA, LinkMap.META_PAGE_IDX);
+
+                assert PageIdUtils.pageIndex(mappingMetaPageId) == LinkMap.META_PAGE_IDX
+                    : PageIdUtils.toDetailString(mappingMetaPageId);
+
+                linkMap = new LinkMap(newGrpCtx, mappingPageMemory, mappingMetaPageId, initNew);
+            }
+            finally {
+                defragmentationCheckpoint.checkpointTimeoutLock().checkpointReadUnlock();
+            }
+
+            return linkMap;
+        }
+
+        /** */
+        public void createNewCacheDataStore(GridSpinBusyLock busyLock) {
```

Comment:
From my understanding 'context' means 'sharing state/references'. For instance, I haven't seen any such fabric methods inside `IgniteKernalContext`. I think we should choose the right place for such actions - OffheapManager seems to be the right place.

Reply:
I think we should have defragmentation code separately. We already have issues with classes responsibility scopes, I don't want to make it worse. I'd like to be as far from OffheapManager as possible.

Reply:
I agree that defragmentation code must be as self-sufficient as possible. But here you are using the `busyLock` from `OffheapManager` as a method parameter and it seems there can be nothing worse than this.
Let's move this factory method to the OffheapManager.

Reply:
I agree with you, I'll create that method